python-checks/src/main/java/org/sonar/python/checks/tests/ImplicitlySkippedTestCheck.java

```diff
@@ -0,0 +1,187 @@
+      "assertNotIn", "assertIsInstance", "assertNotIsInstance", "assertRaises", "assertRaisesRegexp", "assertAlmostEqual",
+      "assertNotAlmostEqual", "assertGreater", "assertGreaterEqual", "assertLess", "assertLessEqual", "assertRegexpMatches",
+      "assertNotRegexpMatches", "assertItemsEqual", "assertDictContainsSubset", "assertMultiLineEqual", "assertSequenceEqual",
+      "assertListEqual", "assertTupleEqual", "assertSetEqual", "assertDictEqual"))));
+  }
+
+  private static final Tree.Kind[] literalsKind = {Tree.Kind.STRING_LITERAL, Tree.Kind.NUMERIC_LITERAL, Tree.Kind.LIST_LITERAL,
+    Tree.Kind.BOOLEAN_LITERAL_PATTERN, Tree.Kind.NUMERIC_LITERAL_PATTERN, Tree.Kind.NONE_LITERAL_PATTERN, Tree.Kind.STRING_LITERAL_PATTERN};
+
+  static class TestFramework {
+    String name;
+    List<String> keywords;
+    Set<String> supportedAssertMethods;
+
+    public TestFramework(String name, List<String> keywords, Set<String> supportedAssertMethods) {
+      this.name = name;
+      this.keywords = keywords;
+      this.supportedAssertMethods = supportedAssertMethods;
+    }
+
+    public boolean matchAnyProvidedClasses(List<String> classes) {
+      return classes.stream().anyMatch(parentClass -> keywords.stream().allMatch(parentClass::contains));
+    }
+  }
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.FUNCDEF, ctx -> {
+      FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+
+      if (!functionDef.name().name().startsWith("test")) {
+        return;
+      }
+
+      ReturnStatement returnStatement = getReturnStatementFromFirstIfStatement(functionDef);
+      if (returnStatement == null) {
+        return;
+      }
+
+      // We add an issue only if there is an assert statement somewhere in the function definition
+      if (containsAssertion(functionDef)) {
+        ctx.addIssue(returnStatement, MESSAGE);
+      }
+    });
+  }
+
+  private static ReturnStatement getReturnStatementFromFirstIfStatement(FunctionDef functionDef) {
+    // check first non assignment statement is an if with a return
+    for (Statement statement : functionDef.body().statements()) {
+      // skip literal assignment expression
+      if (statement.is(Tree.Kind.ASSIGNMENT_STMT) && ((AssignmentStatement) statement).assignedValue().is(literalsKind)) {
+        continue;
+      }
+
+      if (statement.is(Tree.Kind.IF_STMT)) {
+        IfStatement ifStatement = (IfStatement) statement;
+        List<Statement> statements = ifStatement.body().statements();
+        if (statements.get(0).is(Tree.Kind.RETURN_STMT)) {
+          return (ReturnStatement) statements.get(0);
+        }
+      }
+
+      return null;
+    }
+    return null;
+  }
+
+  private static boolean containsAssertion(FunctionDef functionDef) {
+    List<String> parentClasses = getParentClassFromFunctionDef(functionDef);
+    AssertionVisitor assertVisitor = new AssertionVisitor(parentClasses);
```

Comment:
_Minor suggestion:_ I feel we can simplify this a bit by making the assumption that only one test framework is used at once (and that we probably won't add support for many test frameworks in the near future).

We could check before invoking the visitor that the parent classes contain `unittest` and `TestCase`. And if yes, we can provide the list of assertion methods to the visitor (if we are not in a `unittest.TestCase` or derivative, then we provide an empty list).

I'm okay with keeping the current implementation though, this is just a way I think this could be simplified.

Reply:
I adapted a bit, I kept the part where we don't test specifically for unittest framework as I don't like to have this framework hardcoded in the logic, but I modified the part you mentioned in order to simplify and only pass the set of assert methods to the visitor.